File: core/java/src/main/java/hara/lang/data/Deque.java
package hara.lang.data;

import hara.lang.base.Ex;
import hara.lang.protocol.ILinearType;
import hara.lang.data.types.ISequentialLookupType;
import hara.lang.data.types.ObjPersistent;
import hara.lang.protocol.IAssoc;
import hara.lang.protocol.IMetadata;
import hara.lang.protocol.IPopFirst;
import hara.lang.protocol.IPopLast;
import hara.lang.protocol.IPushFirst;
import hara.lang.protocol.IPushLast;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;

/** Persistent deque backed by a count-measured finger tree. */
public interface Deque<E>
    extends ILinearType<E>,
        ISequentialLookupType<E>,
        IAssoc<Long, E>,
        IPushFirst<E>,
        IPushLast<E>,
        IPopFirst,
        IPopLast {
  abstract class Item<E> {
    abstract long measure();
    abstract E get(long index);
    abstract Item<E> replace(long index, E value);
    abstract void collect(ArrayList<E> output);

    java.util.List<Item<E>> children() {
      throw new IllegalStateException("leaf has no children");
    }
  }

  final class Leaf<E> extends Item<E> {
    final E value;
    Leaf(E value) { this.value = value; }
    long measure() { return 1; }
    E get(long index) { return index == 0 ? value : null; }
    Item<E> replace(long index, E next) { return index == 0 ? new Leaf<>(next) : null; }
    void collect(ArrayList<E> output) { output.add(value); }
  }

  final class Branch<E> extends Item<E> {
    final long measure;
    final java.util.List<Item<E>> children;
    Branch(java.util.List<Item<E>> children) {
      if (children.size() < 2 || children.size() > 3) throw new IllegalArgumentException();
      this.children = immutable(children);
      long total = 0;
      for (Item<E> child : children) total += child.measure();
      this.measure = total;
    }
    long measure() { return measure; }
    E get(long index) {
      for (Item<E> child : children) {
        if (index < child.measure()) return child.get(index);
        index -= child.measure();
      }
      return null;
    }
    Item<E> replace(long index, E value) {
      for (int position = 0; position < children.size(); position++) {
        Item<E> child = children.get(position);
        if (index < child.measure()) {
          Item<E> replacement = child.replace(index, value);
          if (replacement == null) return null;
          ArrayList<Item<E>> next = new ArrayList<>(children);
          next.set(position, replacement);
          return new Branch<>(next);
        }
        index -= child.measure();
      }
      return null;
    }
    void collect(ArrayList<E> output) { for (Item<E> child : children) child.collect(output); }
    java.util.List<Item<E>> children() { return children; }
  }

  abstract class FingerTree<E> {
    abstract long measure();
    abstract FingerTree<E> pushFirst(Item<E> item);
    abstract FingerTree<E> pushLast(Item<E> item);
    abstract View<E> popFirst();
    abstract View<E> popLast();
    abstract E get(long index);
    abstract FingerTree<E> replace(long index, E value);
    abstract void collect(ArrayList<E> output);

    static <E> FingerTree<E> empty() { return new Empty<>(); }
    static <E> FingerTree<E> fromItems(java.util.List<Item<E>> items) {
      FingerTree<E> tree = empty();
      for (Item<E> item : items) tree = tree.pushLast(item);
      return tree;
    }
  }

  final class Empty<E> extends FingerTree<E> {
    long measure() { return 0; }
    FingerTree<E> pushFirst(Item<E> item) { return new Single<>(item); }
    FingerTree<E> pushLast(Item<E> item) { return new Single<>(item); }
    View<E> popFirst() { return null; }
    View<E> popLast() { return null; }
    E get(long index) { return null; }
    FingerTree<E> replace(long index, E value) { return null; }
    void collect(ArrayList<E> output) {}
  }

  final class Single<E> extends FingerTree<E> {
    final Item<E> item;
    Single(Item<E> item) { this.item = item; }
    long measure() { return item.measure(); }
    FingerTree<E> pushFirst(Item<E> next) {
      return new Deep<>(items(next), FingerTree.empty(), items(item));
    }
    FingerTree<E> pushLast(Item<E> next) {
      return new Deep<>(items(item), FingerTree.empty(), items(next));
    }
    View<E> popFirst() { return new View<>(item, FingerTree.empty()); }
    View<E> popLast() { return new View<>(item, FingerTree.empty()); }
    E get(long index) { return item.get(index); }
    FingerTree<E> replace(long index, E value) {
      Item<E> replacement = item.replace(index, value);
      return replacement == null ? null : new Single<>(replacement);
    }
    void collect(ArrayList<E> output) { item.collect(output); }
  }

  final class Deep<E> extends FingerTree<E> {
    final long measure;
    final java.util.List<Item<E>> prefix;
    final FingerTree<E> middle;
    final java.util.List<Item<E>> suffix;
    Deep(java.util.List<Item<E>> prefix, FingerTree<E> middle, java.util.List<Item<E>> suffix) {
      if (prefix.isEmpty() || prefix.size() > 4 || suffix.isEmpty() || suffix.size() > 4)
        throw new IllegalArgumentException();
      this.prefix = immutable(prefix);
      this.middle = middle;
      this.suffix = immutable(suffix);
      long total = middle.measure();
      for (Item<E> item : prefix) total += item.measure();
      for (Item<E> item : suffix) total += item.measure();
      this.measure = total;
    }
    long measure() { return measure; }
    FingerTree<E> pushFirst(Item<E> item) {
      if (prefix.size() < 4) {
        ArrayList<Item<E>> next = new ArrayList<>(); next.add(item); next.addAll(prefix);
        return new Deep<>(next, middle, suffix);
      }
      return new Deep<>(items(item, prefix.get(0)),
          middle.pushFirst(new Branch<>(prefix.subList(1, 4))), suffix);
    }
    FingerTree<E> pushLast(Item<E> item) {
      if (suffix.size() < 4) {
        ArrayList<Item<E>> next = new ArrayList<>(suffix); next.add(item);
        return new Deep<>(prefix, middle, next);
      }
      return new Deep<>(prefix, middle.pushLast(new Branch<>(suffix.subList(0, 3))),
          items(suffix.get(3), item));
    }
    View<E> popFirst() {
      Item<E> first = prefix.get(0);
      if (prefix.size() > 1) return new View<>(first, new Deep<>(prefix.subList(1, prefix.size()), middle, suffix));
      View<E> view = middle.popFirst();
      return view == null
          ? new View<>(first, FingerTree.fromItems(suffix))
          : new View<>(first, new Deep<>(view.item.children(), view.tree, suffix));
    }
    View<E> popLast() {
      Item<E> last = suffix.get(suffix.size() - 1);
      if (suffix.size() > 1) return new View<>(last, new Deep<>(prefix, middle, suffix.subList(0, suffix.size() - 1)));
      View<E> view = middle.popLast();
      return view == null
          ? new View<>(last, FingerTree.fromItems(prefix))
          : new View<>(last, new Deep<>(prefix, view.tree, view.item.children()));
    }
    E get(long index) {
      for (Item<E> item : prefix) { if (index < item.measure()) return item.get(index); index -= item.measure(); }
      if (index < middle.measure()) return middle.get(index);
      index -= middle.measure();
      for (Item<E> item : suffix) { if (index < item.measure()) return item.get(index); index -= item.measure(); }
      return null;
    }
    FingerTree<E> replace(long index, E value) {
      ArrayList<Item<E>> next;
      for (int position = 0; position < prefix.size(); position++) {
        Item<E> item = prefix.get(position);
        if (index < item.measure()) { next = new ArrayList<>(prefix); next.set(position, item.replace(index, value)); return new Deep<>(next, middle, suffix); }
        index -= item.measure();
      }
      if (index < middle.measure()) return new Deep<>(prefix, middle.replace(index, value), suffix);
      index -= middle.measure();
      for (int position = 0; position < suffix.size(); position++) {
        Item<E> item = suffix.get(position);
        if (index < item.measure()) { next = new ArrayList<>(suffix); next.set(position, item.replace(index, value)); return new Deep<>(prefix, middle, next); }
        index -= item.measure();
      }
      return null;
    }
    void collect(ArrayList<E> output) {
      for (Item<E> item : prefix) item.collect(output);
      middle.collect(output);
      for (Item<E> item : suffix) item.collect(output);
    }
  }

  final class View<E> {
    final Item<E> item;
    final FingerTree<E> tree;
    View(Item<E> item, FingerTree<E> tree) { this.item = item; this.tree = tree; }
  }

  @SafeVarargs
  static <E> java.util.List<Item<E>> items(Item<E>... values) {
    ArrayList<Item<E>> output = new ArrayList<>();
    Collections.addAll(output, values);
    return output;
  }
  static <E> java.util.List<Item<E>> immutable(java.util.List<Item<E>> values) {
    return Collections.unmodifiableList(new ArrayList<>(values));
  }

  final class Standard<E> extends ObjPersistent implements Deque<E> {
    private final FingerTree<E> tree;
    private static final Standard<Object> EMPTY = new Standard<>(null, FingerTree.empty());
    private Standard(IMetadata metadata, FingerTree<E> tree) { super(metadata); this.tree = tree; }
    @SuppressWarnings("unchecked") public static <E> Standard<E> empty(IMetadata metadata) {
      Standard<E> empty = (Standard<E>) EMPTY;
      return metadata == null ? empty : empty.withMeta(metadata);
    }
    @SafeVarargs public static <E> Standard<E> from(IMetadata metadata, E... values) {
      Standard<E> output = empty(metadata);
      for (E value : values) output = output.pushLast(value);
      return output;
    }
    public static <E> Standard<E> into(Iterator<E> values) {
      Standard<E> output = empty(null);
      while (values.hasNext()) output = output.pushLast(values.next());
      return output;
    }
    public long count() { return tree.measure(); }
    public E nth(long index) {
      if (index < 0 || index >= count()) throw new Ex.OutOfBounds();
      return tree.get(index);
    }
    public Iterator<E> iterator() { ArrayList<E> values = new ArrayList<>(); tree.collect(values); return values.iterator(); }
    public Standard<E> pushFirst(E value) { return new Standard<>(_meta, tree.pushFirst(new Leaf<>(value))); }
    public Standard<E> pushLast(E value) { return new Standard<>(_meta, tree.pushLast(new Leaf<>(value))); }
    public Standard<E> popFirst() { View<E> view = tree.popFirst(); return view == null ? this : new Standard<>(_meta, view.tree); }
    public Standard<E> popLast() { View<E> view = tree.popLast(); return view == null ? this : new Standard<>(_meta, view.tree); }
    public Standard<E> assoc(Long index, E value) {
      FingerTree<E> next = index == null ? null : tree.replace(index, value);
      if (next == null) throw new Ex.OutOfBounds();
      return new Standard<>(_meta, next);
    }
    public Standard<E> withMeta(IMetadata metadata) { return metadata == _meta ? this : new Standard<>(metadata, tree); }
    public Standard<E> empty() { return empty(_meta); }
  }
}

File: core/rust/src/lang/data/deque.rs
//! Persistent deque backed by a count-measured finger tree.

use crate::lang::hash::JavaHash;
use crate::lang::protocol::{
    HashType, IAssoc, IColl, IConj, ICons, ICount, IDisplay, IEmpty, IEquality, IHash, ILookup,
    IMetadata, INth, IObjType, IPeekFirst, IPeekLast, IPersistent, IPopFirst, IPopLast, IPushFirst,
    IPushLast, MetaType, ObjType,
};
use std::rc::Rc;

#[derive(Debug, Clone)]
enum Item<E> {
    Leaf(E),
    Branch {
        measure: usize,
        children: Vec<Rc<Item<E>>>,
    },
}

impl<E: Clone> Item<E> {
    fn leaf(value: E) -> Rc<Self> {
        Rc::new(Self::Leaf(value))
    }
    fn branch(children: Vec<Rc<Self>>) -> Rc<Self> {
        debug_assert!((2..=3).contains(&children.len()));
        Rc::new(Self::Branch {
            measure: children.iter().map(|child| child.measure()).sum(),
            children,
        })
    }
    fn measure(&self) -> usize {
        match self {
            Self::Leaf(_) => 1,
            Self::Branch { measure, .. } => *measure,
        }
    }
    fn get(&self, mut index: usize) -> Option<&E> {
        match self {
            Self::Leaf(value) => (index == 0).then_some(value),
            Self::Branch { children, .. } => {
                for child in children {
                    if index < child.measure() {
                        return child.get(index);
                    }
                    index -= child.measure();
                }
                None
            }
        }
    }
    fn replace(&self, mut index: usize, value: E) -> Option<Rc<Self>> {
        match self {
            Self::Leaf(_) => (index == 0).then(|| Self::leaf(value)),
            Self::Branch { children, .. } => {
                for (position, child) in children.iter().enumerate() {
                    if index < child.measure() {
                        let mut replaced = children.clone();
                        replaced[position] = child.replace(index, value)?;
                        return Some(Self::branch(replaced));
                    }
                    index -= child.measure();
                }
                None
            }
        }
    }
    fn collect<'a>(&'a self, output: &mut Vec<&'a E>) {
        match self {
            Self::Leaf(value) => output.push(value),
            Self::Branch { children, .. } => {
                for child in children {
                    child.collect(output);
                }
            }
        }
    }
    fn children(&self) -> Vec<Rc<Self>> {
        match self {
            Self::Branch { children, .. } => children.clone(),
            Self::Leaf(_) => unreachable!("finger-tree middle item must be a branch"),
        }
    }
}

#[derive(Debug, Clone, Default)]
enum FingerTree<E> {
    #[default]
    Empty,
    Single(Rc<Item<E>>),
    Deep {
        measure: usize,
        prefix: Vec<Rc<Item<E>>>,
        middle: Rc<FingerTree<E>>,
        suffix: Vec<Rc<Item<E>>>,
    },
}

impl<E: Clone> FingerTree<E> {
    fn measure(&self) -> usize {
        match self {
            Self::Empty => 0,
            Self::Single(item) => item.measure(),
            Self::Deep { measure, .. } => *measure,
        }
    }
    fn deep(prefix: Vec<Rc<Item<E>>>, middle: Rc<Self>, suffix: Vec<Rc<Item<E>>>) -> Self {
        debug_assert!((1..=4).contains(&prefix.len()));
        debug_assert!((1..=4).contains(&suffix.len()));
        let measure = prefix.iter().map(|item| item.measure()).sum::<usize>()
            + middle.measure()
            + suffix.iter().map(|item| item.measure()).sum::<usize>();
        Self::Deep {
            measure,
            prefix,
            middle,
            suffix,
        }
    }
    fn push_first_item(&self, item: Rc<Item<E>>) -> Self {
        match self {
            Self::Empty => Self::Single(item),
            Self::Single(existing) => {
                Self::deep(vec![item], Rc::new(Self::Empty), vec![existing.clone()])
            }
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } if prefix.len() < 4 => {
                let mut next = Vec::with_capacity(prefix.len() + 1);
                next.push(item);
                next.extend(prefix.iter().cloned());
                Self::deep(next, middle.clone(), suffix.clone())
            }
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } => {
                let node = Item::branch(prefix[1..].to_vec());
                Self::deep(
                    vec![item, prefix[0].clone()],
                    Rc::new(middle.push_first_item(node)),
                    suffix.clone(),
                )
            }
        }
    }
    fn push_last_item(&self, item: Rc<Item<E>>) -> Self {
        match self {
            Self::Empty => Self::Single(item),
            Self::Single(existing) => {
                Self::deep(vec![existing.clone()], Rc::new(Self::Empty), vec![item])
            }
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } if suffix.len() < 4 => {
                let mut next = suffix.clone();
                next.push(item);
                Self::deep(prefix.clone(), middle.clone(), next)
            }
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } => {
                let node = Item::branch(suffix[..3].to_vec());
                Self::deep(
                    prefix.clone(),
                    Rc::new(middle.push_last_item(node)),
                    vec![suffix[3].clone(), item],
                )
            }
        }
    }
    fn from_items(items: Vec<Rc<Item<E>>>) -> Self {
        items
            .into_iter()
            .fold(Self::Empty, |tree, item| tree.push_last_item(item))
    }
    fn pop_first_item(&self) -> Option<(Rc<Item<E>>, Self)> {
        match self {
            Self::Empty => None,
            Self::Single(item) => Some((item.clone(), Self::Empty)),
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } if prefix.len() > 1 => Some((
                prefix[0].clone(),
                Self::deep(prefix[1..].to_vec(), middle.clone(), suffix.clone()),
            )),
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } => {
                if let Some((node, next_middle)) = middle.pop_first_item() {
                    Some((
                        prefix[0].clone(),
                        Self::deep(node.children(), Rc::new(next_middle), suffix.clone()),
                    ))
                } else {
                    Some((prefix[0].clone(), Self::from_items(suffix.clone())))
                }
            }
        }
    }
    fn pop_last_item(&self) -> Option<(Rc<Item<E>>, Self)> {
        match self {
            Self::Empty => None,
            Self::Single(item) => Some((item.clone(), Self::Empty)),
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } if suffix.len() > 1 => {
                let last = suffix.len() - 1;
                Some((
                    suffix[last].clone(),
                    Self::deep(prefix.clone(), middle.clone(), suffix[..last].to_vec()),
                ))
            }
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } => {
                if let Some((node, next_middle)) = middle.pop_last_item() {
                    Some((
                        suffix[0].clone(),
                        Self::deep(prefix.clone(), Rc::new(next_middle), node.children()),
                    ))
                } else {
                    Some((suffix[0].clone(), Self::from_items(prefix.clone())))
                }
            }
        }
    }
    fn get(&self, mut index: usize) -> Option<&E> {
        match self {
            Self::Empty => None,
            Self::Single(item) => item.get(index),
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } => {
                for item in prefix {
                    if index < item.measure() {
                        return item.get(index);
                    }
                    index -= item.measure();
                }
                if index < middle.measure() {
                    return middle.get(index);
                }
                index -= middle.measure();
                for item in suffix {
                    if index < item.measure() {
                        return item.get(index);
                    }
                    index -= item.measure();
                }
                None
            }
        }
    }
    fn replace(&self, mut index: usize, value: E) -> Option<Self> {
        match self {
            Self::Empty => None,
            Self::Single(item) => Some(Self::Single(item.replace(index, value)?)),
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } => {
                for (position, item) in prefix.iter().enumerate() {
                    if index < item.measure() {
                        let mut next = prefix.clone();
                        next[position] = item.replace(index, value)?;
                        return Some(Self::deep(next, middle.clone(), suffix.clone()));
                    }
                    index -= item.measure();
                }
                if index < middle.measure() {
                    return Some(Self::deep(
                        prefix.clone(),
                        Rc::new(middle.replace(index, value)?),
                        suffix.clone(),
                    ));
                }
                index -= middle.measure();
                for (position, item) in suffix.iter().enumerate() {
                    if index < item.measure() {
                        let mut next = suffix.clone();
                        next[position] = item.replace(index, value)?;
                        return Some(Self::deep(prefix.clone(), middle.clone(), next));
                    }
                    index -= item.measure();
                }
                None
            }
        }
    }
    fn collect<'a>(&'a self, output: &mut Vec<&'a E>) {
        match self {
            Self::Empty => {}
            Self::Single(item) => item.collect(output),
            Self::Deep {
                prefix,
                middle,
                suffix,
                ..
            } => {
                for item in prefix {
                    item.collect(output);
                }
                middle.collect(output);
                for item in suffix {
                    item.collect(output);
                }
            }
        }
    }
}

#[derive(Debug, Clone)]
pub struct Standard<E> {
    metadata: Option<Rc<crate::lang::data::Metadata>>,
    tree: FingerTree<E>,
}

impl<E: Clone> Default for Standard<E> {
    fn default() -> Self {
        Self {
            metadata: None,
            tree: FingerTree::Empty,
        }
    }
}
impl<E: Clone> Standard<E> {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn len(&self) -> usize {
        self.tree.measure()
    }
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
    pub fn get(&self, index: usize) -> Option<&E> {
        self.tree.get(index)
    }
    pub fn push_first(&self, value: E) -> Self {
        Self {
            metadata: self.metadata.clone(),
            tree: self.tree.push_first_item(Item::leaf(value)),
        }
    }
    pub fn push_last(&self, value: E) -> Self {
        Self {
            metadata: self.metadata.clone(),
            tree: self.tree.push_last_item(Item::leaf(value)),
        }
    }
    pub fn pop_first_value(&self) -> Self {
        self.tree
            .pop_first_item()
            .map(|(_, tree)| Self {
                metadata: self.metadata.clone(),
                tree,
            })
            .unwrap_or_else(|| self.clone())
    }
    pub fn pop_last_value(&self) -> Self {
        self.tree
            .pop_last_item()
            .map(|(_, tree)| Self {
                metadata: self.metadata.clone(),
                tree,
            })
            .unwrap_or_else(|| self.clone())
    }
    pub fn assoc_value(&self, index: usize, value: E) -> Option<Self> {
        Some(Self {
            metadata: self.metadata.clone(),
            tree: self.tree.replace(index, value)?,
        })
    }
    pub fn iter(&self) -> std::vec::IntoIter<&E> {
        let mut values = Vec::with_capacity(self.len());
        self.tree.collect(&mut values);
        values.into_iter()
    }
}
impl<E: Clone> FromIterator<E> for Standard<E> {
    fn from_iter<T: IntoIterator<Item = E>>(it: T) -> Self {
        it.into_iter().fold(Self::new(), |d, v| d.push_last(v))
    }
}
impl<E: Clone> IntoIterator for Standard<E> {
    type Item = E;
    type IntoIter = std::vec::IntoIter<E>;
    fn into_iter(self) -> Self::IntoIter {
        self.iter().cloned().collect::<Vec<_>>().into_iter()
    }
}
impl<E: Clone + PartialEq> PartialEq for Standard<E> {
    fn eq(&self, other: &Self) -> bool {
        self.len() == other.len() && self.iter().eq(other.iter())
    }
}
impl<E: Clone> ICount for Standard<E> {
    fn count(&self) -> usize {
        self.len()
    }
}
impl<E: Clone> INth<E> for Standard<E> {
    fn nth(&self, index: usize) -> Option<&E> {
        self.get(index)
    }
}
impl<E: Clone> ILookup<usize, E> for Standard<E> {
    type Keys = std::ops::Range<usize>;
    type Values = std::vec::IntoIter<E>;
    fn keys(&self) -> Self::Keys {
        0..self.len()
    }
    fn vals(&self) -> Self::Values {
        self.iter().cloned().collect::<Vec<_>>().into_iter()
    }
}
impl<E: Clone> crate::lang::protocol::IFind<usize> for Standard<E> {
    type Output = (usize, E);
    fn find(&self, key: &usize) -> Option<Self::Output> {
        self.get(*key).cloned().map(|v| (*key, v))
    }
}
impl<E: Clone> IAssoc<usize, E> for Standard<E> {
    type Output = Self;
    fn assoc(&self, k: usize, v: E) -> Self {
        self.assoc_value(k, v).unwrap_or_else(|| self.clone())
    }
}
impl<E: Clone> IPeekFirst<E> for Standard<E> {
    fn peek_first(&self) -> Option<E> {
        self.get(0).cloned()
    }
}
impl<E: Clone> IPeekLast<E> for Standard<E> {
    fn peek_last(&self) -> Option<E> {
        self.len().checked_sub(1).and_then(|i| self.get(i)).cloned()
    }
}
impl<E: Clone> IPushFirst<E> for Standard<E> {
    type Output = Self;
    fn push_first(&self, v: E) -> Self {
        Standard::push_first(self, v)
    }
}
impl<E: Clone> IPushLast<E> for Standard<E> {
    type Output = Self;
    fn push_last(&self, v: E) -> Self {
        Standard::push_last(self, v)
    }
}
impl<E: Clone> IPopFirst for Standard<E> {
    type Output = Self;
    fn pop_first(&self) -> Self {
        self.pop_first_value()
    }
}
impl<E: Clone> IPopLast for Standard<E> {
    type Output = Self;
    fn pop_last(&self) -> Self {
        self.pop_last_value()
    }
}
impl<E: Clone> ICons<E> for Standard<E> {
    type Output = Self;
    fn cons(&self, v: E) -> Self {
        self.push_first(v)
    }
}
impl<E: Clone> IConj<E> for Standard<E> {
    type Output = Self;
    fn conj(&self, v: E) -> Self {
        self.push_last(v)
    }
}
impl<E: Clone> IEmpty for Standard<E> {
    type Output = Self;
    fn empty(&self) -> Self {
        Self::new().with_meta(self.metadata.clone())
    }
}
impl<E: Clone> IMetadata for Standard<E> {
    type Metadata = Rc<crate::lang::data::Metadata>;
    fn meta(&self) -> Option<&Self::Metadata> {
        self.metadata.as_ref()
    }
    fn with_meta(&self, metadata: Option<Self::Metadata>) -> Self {
        Self {
            metadata,
            tree: self.tree.clone(),
        }
    }
    fn metatype(&self) -> MetaType {
        MetaType::Object
    }
}
impl<E: Clone> IPersistent for Standard<E> {}
impl<E: Clone + PartialEq> IEquality for Standard<E> {
    fn equality(&self, other: &Self) -> bool {
        self == other
    }
}
impl<E: Clone + std::fmt::Debug> IDisplay for Standard<E> {
    fn display(&self) -> String {
        format!(
            "[{}]",
            self.iter()
                .map(|v| format!("{v:?}"))
                .collect::<Vec<_>>()
                .join(" ")
        )
    }
}
impl<E: Clone + std::hash::Hash + JavaHash> IHash for Standard<E> {
    fn hash_calc(&self, t: HashType) -> u64 {
        crate::lang::hash::compose_ordered("SEQUENTIAL", self.iter().map(|v| v.java_hash(t))) as u64
    }
}
impl<E: Clone + std::fmt::Debug> IObjType for Standard<E> {
    fn obj_type(&self) -> ObjType {
        ObjType::Sequential
    }
}
impl<E> IColl<E> for Standard<E>
where
    E: Clone + PartialEq + std::hash::Hash + JavaHash + std::fmt::Debug,
{
    fn start_string(&self) -> &'static str {
        "["
    }
    fn end_string(&self) -> &'static str {
        "]"
    }
}

#[cfg(test)]
mod tests {
    use super::Standard;
    #[test]
    fn finger_tree_deque_matches_two_ended_model() {
        let mut deque = Standard::new();
        let mut model = std::collections::VecDeque::new();
        for i in 0..2000 {
            if i % 3 == 0 {
                deque = deque.push_first(i);
                model.push_front(i)
            } else {
                deque = deque.push_last(i);
                model.push_back(i)
            }
        }
        assert_eq!(
            deque.iter().copied().collect::<Vec<_>>(),
            model.iter().copied().collect::<Vec<_>>()
        );
        for i in 0..1500 {
            if i % 2 == 0 {
                deque = deque.pop_first_value();
                model.pop_front();
            } else {
                deque = deque.pop_last_value();
                model.pop_back();
            }
        }
        assert_eq!(
            deque.iter().copied().collect::<Vec<_>>(),
            model.iter().copied().collect::<Vec<_>>()
        );
        assert_eq!(deque.len(), model.len());
    }
    #[test]
    fn persistence_and_index_replacement_are_preserved() {
        let original: Standard<_> = (0..100).collect();
        let changed = original.assoc_value(50, 999).unwrap();
        assert_eq!(original.get(50), Some(&50));
        assert_eq!(changed.get(50), Some(&999));
        assert_eq!(changed.get(99), Some(&99));
    }
}
